bundles/nl.asml.matala.product/src/nl/asml/matala/product/validation/ProductValidator.xtend

```diff
@@ -30,6 +49,115 @@ class ProductValidator extends AbstractProductValidator {
+        var HashSet<String> existingLocalvars = new HashSet<String>();
+            for (localvars : b.localvars) {
+                if (!existingLocalvars.add(localvars.name)) {
+                    error("There exists a variable name in local with the same name" ,  ProductPackage.Literals.BLOCK__LOCALVARS)  
+                }
+            }
+        }
+
+    
+    
+        
+    /**
+     * All variables occurring in a guard expression must have been defined as inputs of an action
+     */
+    @Check
+    def preventIlligalVariableAccess(Function f){
+          var allVariables = new HashSet<String>
+          var inputes = new HashSet<String>
+          for(update : f.updates){
+            if(update.guard !== null){
+               // Get all variables in a guard
+                val allExprVariables = EcoreUtil2::getAllContentsOfType(update.guard, ExpressionVariable)
+                for(v : allExprVariables){
+                        allVariables.add( v.variable.name)
+                   }
+                //get all the variables in input
+                 for (element : update.fnInp) {
+                        inputes.add(element.ref.name)                      
+                    } 
+                if (!inputes.containsAll(allVariables)){
+                    error("The variable used in guard is not defined in the input variables" , ProductPackage.Literals.FUNCTION__UPDATES)
+                }
+            }
+        }
+    }
+
+    /**
+     * Variables occurring in the RHS of an update expression must have been defined as inputs of an action.
+     *  LHS variable is the same as that mentioned in immediately preceding "updates". Note that attributes of this variable may be referenced in the LHS expression (record field access)
+     */
+    @Check
+    def preventIlligalVariableAccessUpdate(Function F){
+         /* Variables occurring in the RHS of an update expression must have been defined as inputs of an action. */
+           for(update : F.updates){               
+               var inputes = new HashSet<String>               
+               //get with-inputs of an action
+               for(fnInp : update.fnInp){
+                   inputes.add(fnInp.ref.name)                  
+               }  
+             for(updateOutVar :  update.updateOutputVar){
+                 var producesOutputs = new HashSet<String>
+                 for(fnOut : updateOutVar.fnOut){
+                       producesOutputs.add(fnOut.ref.name)                          
+                   }
+                 var actions = updateOutVar.act.actions
+                 var allVariables = new HashSet<String>                     
+                 for(action : actions){                      
+                    val helperAction = ValidationHelper.action(action)                              
+                    val LHSValue = helperAction.get("LHS") as String
+                    val expValue = helperAction.get("RHS") as Expression
```

Comment:
If the 'RHS' is always an `Expression`, why not change it in the API of the function?